lib/livebook_web/live/session_live/live_widget/vega_lite_live.ex

```diff
@@ -0,0 +1,32 @@
+defmodule LivebookWeb.SessionLive.LiveWidget.VegaLiteLive do
+  use LivebookWeb, :live_view
+
+  @impl true
+  def mount(_params, %{"pid" => pid, "id" => id}, socket) do
+    send(pid, {:connect, self()})
```

Comment:
I assume that we "subscribe" to push events on the connect call, correct?

In this case, the approach you have today has a race condition. If we call LiveWidget.render and then immediately start pushing events, we may not have subscribed yet. So by the time you start pushing events, some of them may have been missed.

You could solve it if you made the "render" operation a sync operation but that adds some complexity. Is this an issue? Do you have any other idea how to address it besides removing LiveWidget.render?

Reply:
Well, alternatively the runtime can keep a copy of the whole data set and push it on subscription... but that is bound to cause problems too (for this use case).

Reply:
I think this is similar to having the widget already on the page and the refreshing it (or another user joining). What I did for VegaLite is that the process keeps the data (within the specified window, if any), so [after connecting](https://github.com/jonatanklosko/live_widget/blob/60f9cacd9a7bbb7f0add9773f570056b8d554a6f/lib/live_widget/vega_lite.ex#L152-L160) the data is sent to the subscriber.